src/Npgsql/NpgsqlRecordDataReader.cs

```diff
@@ -0,0 +1,327 @@
+        {
+            var dimensions = Buffer.ReadInt32();
+            var containsNulls = Buffer.ReadInt32() == 1;
+            Buffer.ReadUInt32(); // Element OID. Ignored.
+
+            if (containsNulls)
+                throw new InvalidOperationException("Record array contains null record");
+
+            if (dimensions == 0)
+                return;
+
+            if (dimensions != 1)
+                throw new InvalidOperationException($"Cannot read an array with 1 dimension from an array with {dimensions} dimension(s)");
+
+            _numRows = Buffer.ReadInt32();
+            Buffer.ReadInt32(); // Lower bound
+
+            if (_numRows > 0)
+                Buffer.ReadInt32(); // Length of first row
+
+            _nextRowBufferPos = Buffer.ReadPosition;
+        }
+
+        internal void InitSingleRow()
+        {
+            _numRows = 1;
+            _nextRowBufferPos = Buffer.ReadPosition;
+        }
+
+        public override object this[int ordinal] => GetValue(ordinal);
+
+        public override object this[string name] => GetValue(GetOrdinal(name));
+
+        public override int Depth
+        {
+            get
+            {
+                CheckNotClosed();
+                return _depth;
+            }
+        }
+
+        public override int FieldCount
+        {
+            get
+            {
+                CheckNotClosed();
+                return _calledNextResult ? 0 : _columns?.Length ?? 0;
+            }
+        }
+
+        public override bool HasRows
+        {
+            get
+            {
+                CheckNotClosed();
+                return _numRows > 0;
+            }
+        }
+
+        public override bool IsClosed => _closed;
+
+        public override int RecordsAffected => -1;
+
+        public override bool GetBoolean(int ordinal) => GetFieldValue<bool>(ordinal);
+        public override byte GetByte(int ordinal) => GetFieldValue<byte>(ordinal);
+        public override char GetChar(int ordinal) => GetFieldValue<char>(ordinal);
+        public override DateTime GetDateTime(int ordinal) => GetFieldValue<DateTime>(ordinal);
+        public override decimal GetDecimal(int ordinal) => GetFieldValue<decimal>(ordinal);
+        public override double GetDouble(int ordinal) => GetFieldValue<double>(ordinal);
+        public override float GetFloat(int ordinal) => GetFieldValue<float>(ordinal);
+        public override Guid GetGuid(int ordinal) => GetFieldValue<Guid>(ordinal);
+        public override short GetInt16(int ordinal) => GetFieldValue<short>(ordinal);
+        public override int GetInt32(int ordinal) => GetFieldValue<int>(ordinal);
+        public override long GetInt64(int ordinal) => GetFieldValue<long>(ordinal);
+        public override string GetString(int ordinal) => GetFieldValue<string>(ordinal);
+        public override long GetBytes(int ordinal, long dataOffset, byte[]? buffer, int bufferOffset, int length)
+            => throw new NotImplementedException();
```

Comment:
Change to NotSupportedException, assuming we don't intend to actually implement this for now (I'd wait until someone asks for it).

Reply:
Didn't bother about this one for now in this initial version...

Reply:
Agree, so better to throw NotSupportedException.